### ISSUE TITLE
Fix @Cleanup

### ISSUE THREAD

**Issue by lombokissues** [projectlombok/lombok#834]
_Migrated from Google Code (issue 799)_

**Comment by einar-saukas** [comment]
Bump!

I'm starting to suspect this issue has been forgotten, It's supposedly a simple change, that will fix a problem relevant enough to be described in the main documentation page itself. Thus I'm surprised this issue is still not assigned almost a year later :(

**Comment by almson** [comment]
I don't know when this issue was written, or when the lombok docs were last implemented, but ever since try-with-resources, Java has offered **suppressed exceptions**.

It's all explained here: http://www.oracle.com/technetwork/articles/java/trywithresources-401775.html#nb2

I think `@Cleanup` should work the same way as try-with-resources.

**Comment by einar-saukas** [comment]
Good point! I agree @Cleanup should preserve suppressed exceptions whenever possible.

In practice, it won't make much difference. Storing suppressed exceptions in Throwable is only available in Java 7 or later, but @Cleanup is only really useful prior to Java 7 (before try-with-resources became available). Even so, making @Cleanup fully compatible with try-with-resources in Java 7 or later is an excellent idea.

In this case, the original example:

```
public static void main(String[] args) throws IOException {
    @Cleanup InputStream in = new FileInputStream(args[0]);
    in.read();
    ...
}

```
would be better expanded in Lombok as follows:

```
public static void main(String[] args) throws IOException {
    InputStream in = new FileInputStream(args[0]);
    try {
        in.read();
        ...
        if (in != null) {
            try {
                in.close();
            } finally {
                in = null;
            }
        }
    } catch (Throwable t) {
        if (in != null) {
            try {
                in.close();
            } catch (Throwable e) {
                // t.addSuppressed(e);
                try {
                    Throwable.class.getMethod("addSuppressed", new Class[] { Throwable.class }).invoke(t, e);
                } catch (Exception e2) {
                    // ignore NoSuchMethodException, etc.
                }
            }
        }
        throw Lombok.sneakyThrow(t);
    }
}

```

**Comment by almson** [comment]
Thanks. People seem to forget about suppressed exceptions. Regarding the implementation, I'm fine with reflection, but you can't put `close` in the body because it will be skipped by `return`. The article I linked to shows Oracle's implementation.

**Comment by almson** [comment]
The article contains an interesting point:
> It also advises not to [throw inside `close()`] any exception that should not be suppressed, with `java.lang.InterruptedException` being the best example. 

Might we improve over try-with-resources by not suppressing InterruptedExceptions?

**Comment by einar-saukas** [comment]
Good point again! I didn't realize `return` would break previous implementation. Thanks a lot for pointing this out!

A more appropriate implementation based on Oracle's example, also taking into account arbitrary exceptions and also validating if the resource wasn't set to null, should take the original example:

```
public static void main(String[] args) throws IOException {
    @Cleanup InputStream in = new FileInputStream(args[0]);
    in.read();
    ...
}
```

then expand it in Lombok as follows:

```
public static void main(String[] args) throws IOException {
    InputStream in = new FileInputStream(args[0]);
    Throwable $t = null;
    try {
        in.read();
        ...
    } catch (Throwable t) {
        $t = t;
        throw Lombok.sneakyThrow(t);
    } finally {
        if (in != null) {
            if ($t != null) {
                try {
                    in.close();
                } catch (Throwable e) {
                    // $t.addSuppressed(e);
                    try {
                        Throwable.class.getMethod("addSuppressed", new Class[] { Throwable.class }).invoke($t, e);
                    } catch (Throwable e2) {
                        // ignore NoSuchMethodException, etc. (in case Java < 1.7)
                    }
                }
            } else {
                in.close();
            }
        }
    }
}
```

I only disagree about providing special treatment for InterruptedExceptions. Perhaps this could be somewhat convenient, but it would make this functionality have an inconsistent non-intuitive behaviour, that would differ from try-with-resources. Inconsistent behaviours make it harder for programmers to use and test their code properly in order to avoid unwanted side-effects. The drawbacks would surpass the benefits in this case, IMHO.

**Comment by almson** [comment]
Looks correct. I understand where you're coming from with keeping the behavior consistent. But that warning from *Oracle* that you shouldn't use try-with-resources together with certain close methods which throw certain kinds of exceptions gave me pause. The logic makes sense. 

Here's the gist of it:

Ultimately, it doesn't *really* matter which exception is suppressed. Both have been thrown. Both have been declared. Both might be handled by an upstream exception handler. And it's highly unlikely those upstream handlers are complex enough to have special behavior for when both are thrown at the same time. What I think should happen is that the **worse** exception should get priority. Ie, we should go up the call stack as much as possible. Therefore, I would not suppress `Error`s, as well as `InterruptedException` (since it's a rare exception that's usually associated these days with forceful cancellation of tasks and should be treated as a non-recoverable Error.)

**Comment by almson** [comment]
Ideally, you could actually determine which exception is handled furthest up the stack. Perhaps some magic with proxies?

**Comment by Maaartinus** [comment]
@einar-saukas  You forgot to throw `$t`.

@almson
> What I think should happen is that the **worse** exception should get priority

This is an interesting point, but very subjective. And it'd hardly ever get used. How often it happens that

- some exception gets thrown in the body
- some *different and more important* exception gets thrown when closing
- there are *different* handlers for the two exceptions

and the handlers do really different things and everything works fine only when your approach gets used?

>  you could actually determine which exception is handled furthest up the stack.

I'm afraid, it's not possible at all. Moreover, you'd have to consider exception wrapping, too.

I'm sure, it's not possible for Lombok. It works in a place, where it's hard to find out the fully qualified class name (search for "resolution") and this is a few orders of magnitudes harder.

**Comment by einar-saukas** [comment]
> Ultimately, it doesn't really matter which exception is suppressed.

I disagree. By convention, the original exception (the "cause") should be kept, and the following exception (the "consequence") should be supressed.

According to the Oracle article you mentioned yourself: "Developers lose significant time debugging when a main exception is masked by a further exception being thrown while closing a resource."

> Both have been thrown. Both have been declared. Both might be handled by an upstream exception handler. And it's highly unlikely those upstream handlers are complex enough to have special behavior for when both are thrown at the same time. What I think should happen is that the worse exception should get priority. Ie, we should go up the call stack as much as possible. Therefore, I would not suppress Errors, as well as InterruptedException (since it's a rare exception that's usually associated these days with forceful cancellation of tasks and should be treated as a non-recoverable Error.)

You are suggesting to keep the "most important" exception, instead of keeping the causal exception. The problem is, exceptions don't have priority, thus the concept of "most important" exception is subjective. You propose that Error should be more important than InterruptedException, which should be more important than other exceptions. According to the same logic, one could except that RuntimeException should be also more important than other exceptions, which should be more important than other throwables. Also a RuntimeErrorException should be more important than a RuntimeException, and so on...

Again, there's also the problem that your proposal would make @Cleanup behaviour more complex, inconsistent, non-intuitive, and differ from try-with-resources.

> Ideally, you could actually determine which exception is handled furthest up the stack. Perhaps some magic with proxies?

Are you suggesting that the same routine using @Cleanup should throw errors in different ways depending on the rest of the program using it? It would become impractical to write proper test cases or even reuse libraries reliably.

**Comment by einar-saukas** [comment]
> @einar-saukas You forgot to throw $t.

No, there's no need.

The first catch block already rethrows t, nobody else catches it again.

**Comment by almson** [comment]
The example that is in my head is that I launch a FutureTask (ie, thread) which catches and handles some exceptions, and lets others propagate up and kill it. For example, the cancellation mechanism for tasks uses `thread.interrupt` and requires that InterruptedException is not caught.

Consider that in the following code, the download throws a CommunicationException which normally causes the download to be re-tried. However, while waiting to close CommunicationChannel, the thread receives an interrupt and tries to throw an InterruptedException. However, this exception is suppressed and the cancellation does not happen.

	FutureTask task = new FutureTask (this::task);
        executorService.submit (this::task);

	...

	task.cancel(); // task thread will be interrupted, throwing an InterruptedException
	// the InterruptedException must not be caught, or the task will not cancel


	int task() throws InterruptedException
	{
		boolean retry = true;
		while (retry)
		{
			try (CommunicationChannel channel = ...)
			{
				// perform download
                                // fail and throw a CommunicationException

				retry = false;
			}
			catch (CommuncationException e)
			{
				// ignore exception and retry
			}
			finally
			{
				// channel.close() will wait and can throw an InterruptedException
                                // which may be suppressed
			}
		}

		return 0;
	}

Hence I said that when two exceptions are thrown, the path of execution should return as far up the stack as possible.

**Comment by einar-saukas** [comment]
Your example goes directly against the article you mentioned yourself, namely:

> The specification for auto-closeable classes [...] also advises not to declare any exception that should not be suppressed, with java.lang.InterruptedException being the best example. 

Although special treatment could be convenient for certain situations as your example, it would make @Cleanup harder to use, and inconsistent with try-with-resources. Thus it's not worth it.

IMHO the only acceptable way to prioritize certain exceptions would be declaring them explicitly as an optional list, such as:

```
@Cleanup(precedence={ InterruptedException.class, RuntimeException.class })
InputStream in = new FileInputStream(args[0]);
```

In this example, if both main block and close block both throw exceptions, then close exception will be still suppressed by default, except if close throws InterruptedException but main block doesn't, or if close throws RuntimeException but main block doesn't throw either InterruptedException or RuntimeException.

**Comment by almson** [comment]
I think the explicit prioritization is better than nothing. Note that because @Cleanup allows calling methods other than AutoCloseable.close, the called method might not obey the AutoCloseable.close contract which does "forbid" InterruptedException. (Actually, the javadoc "strongly advises" not to throw it... does that count as being part of the contract?)

I propose calling the parameter "dontSuppress".

**Comment by einar-saukas** [comment]
Ping!

Current "small print" documentation for @ Cleanup still contains the following remark:

"The authors of lombok do not currently know of a feasible way to implement this scheme, but if java updates allow it, or we find a way, we'll fix it.".

However a way was found over 2 years ago, but it hasn't been fixed yet. Was it forgotten?

**Comment by pedrolamarao** [comment]
I came here while looking for a discussion on a variant @Cleanup specific for AutoCloseable types. Above, it was remarked that try-with-resources makes @Cleanup obsolete. I disagree: @Cleanup brings a lot of legibility which is particularly useful when writing small unit tests. I think it would be great to have some kind of @AutoClose as an improved @Cleanup for Java 7 AutoCloseables.

**Comment by almson** [comment]
Please align @Cleanup with try-with-resources! Actually, just emit try-with-resources. (Why would someone use this? Because try-with-resources requires annoying indentation.) Or, idk, delete it.

Java 1-7 have lost nearly all vendor support. New versions of Lombok shouldn't cater to them!

**Comment by almson** [comment]
@einar-saukas What kind of class has close but isn't AutoCloseable? That's a bit of an abomination.

**Comment by einar-saukas** [comment]
> @einar-saukas What kind of class has close but isn't AutoCloseable?

Older libraries.

Annotation @Cleanup works even for a method to close resources that's not called "close".
